src/cachemachine/cachechecker.py

```diff
@@ -0,0 +1,70 @@
+from collections import defaultdict
+from typing import Dict, List, Set
+
+import structlog
+from kubernetes.client.api import core_v1_api
+
+logger = structlog.get_logger(__name__)
+
+
+class CacheChecker:
+    common_cache: Set[str]
+    labels: Dict[str, str]
+    nodes_exist: bool
+    recommended_names: Dict[str, Set[str]]
+    recommended_image_urls: List[str]
+
+    def __init__(
+        self, labels: Dict[str, str], recommended_image_urls: List[str]
+    ):
+        self.api = core_v1_api.CoreV1Api()
+        self.labels = labels
+        self.recommended_image_urls = recommended_image_urls
+        self._reset()
+
+    def _reset(self) -> None:
+        self.common_cache = set()
+        self.nodes_exist = False
+        self.recommended_names = defaultdict(set)
+
+    def check(self) -> None:
+        self._reset()
+
+        nodes = self.api.list_node().items
+        logger.debug(f"Inspecting {nodes}")
+
+        for n in nodes:
+            node_name = n.metadata.name
+
+            # This is a bit tricky.  The images is a list,
+            # each item containing a particular image, and containing
+            # a list of all the names it is known by.
+            images = []
+            recommended_names: Dict[str, Set[str]] = defaultdict(set)
+            for i in n.status.images:
+                for url in self.recommended_image_urls:
+                    if url in i.names:
+                        recommended_names[url] |= set(i.names)
+
+                images.extend(i.names)
+
+            logger.debug(f"{node_name} images: {images}")
+            logger.debug(f"{node_name} recommended: {recommended_names}")
+
+            # If the labels we have are a subset of the labels
+            # of the node, include those images.
```

Comment:
The labels is all about figuring out which nodes to operate on.  This is just bad comment wording.

If the labels we are _filtering on_ are a subset of the labels of the node, keep track of those images on that node.  Otherwise ignore everything on that node.

Reply:
Ah, that makes sense now, thanks!